File: src/entities/Sensor.cpp
/*
 * Copyright (C) 2006-2018 Christopho, Solarus - http://www.solarus-games.org
 *
 * Solarus is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Solarus is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */
#include "solarus/core/Debug.h"
#include "solarus/core/Game.h"
#include "solarus/core/Map.h"
#include "solarus/core/QuestFiles.h"
#include "solarus/entities/Entities.h"
#include "solarus/entities/Hero.h"
#include "solarus/entities/Sensor.h"
#include "solarus/lua/LuaContext.h"

namespace Solarus {

/**
 * \brief Constructor.
 * \param name name of the entity
 * \param layer layer of the entity
 * \param xy Coordinates where to create the entity.
 * \param size Size of the entity.
 */
Sensor::Sensor(
    const std::string& name,
    int layer,
    const Point& xy,
    const Size& size):
  Entity(name, 0, layer, xy, size),
  activated_by_hero(false),
  notifying_script(false) {

  set_collision_modes(CollisionMode::COLLISION_CONTAINING | CollisionMode::COLLISION_OVERLAPPING);
  set_origin(8, 13);
}

/**
 * \brief Returns the type of entity.
 * \return the type of entity
 */
EntityType Sensor::get_type() const {
  return ThisType;
}

/**
 * \brief Returns whether entities of this type can be drawn.
 * \return true if this type of entity can be drawn
 */
bool Sensor::can_be_drawn() const {
  return false;
}

/**
 * \brief Returns whether this entity is an obstacle for another one.
 * \param other another entity
 * \return true if this entity is an obstacle for the other one
 */
bool Sensor::is_obstacle_for(Entity& other) {

  return other.is_sensor_obstacle(*this);
}

/**
 * \brief This function is called by the engine when an entity overlaps this sensor.
 * \param entity_overlapping the entity overlapping the detector
 * \param collision_mode the collision mode that detected the collision
 */
void Sensor::notify_collision(Entity& entity_overlapping, CollisionMode collision_mode) {

  entity_overlapping.notify_collision_with_sensor(*this, collision_mode);
}

/**
 * \brief This function is called when an explosion detects a collision with this entity.
 * \param explosion an explosion
 * \param collision_mode the collision mode that detected the collision
 */
void Sensor::notify_collision_with_explosion(Explosion& /* explosion */, CollisionMode collision_mode) {

  if (collision_mode == COLLISION_OVERLAPPING) {
    get_lua_context()->sensor_on_collision_explosion(*this);
  }
}

/**
 * \brief Activates this sensor.
 *
 * This function is called when the hero overlaps the sensor.
 *
 * \param hero the hero
 */
void Sensor::activate(Hero& /* hero */) {

  if (!activated_by_hero) {

    activated_by_hero = true;

    // Notify Lua.
    notifying_script = true;
    get_lua_context()->sensor_on_activated(*this);
    notifying_script = false;
  }
  else {
    if (!notifying_script && !get_game().is_suspended()) {
      notifying_script = true;
      get_lua_context()->sensor_on_activated_repeat(*this);
      notifying_script = false;
    }
  }
}

/**
 * \brief Updates this entity.
 */
void Sensor::update() {

  Entity::update();

  if (activated_by_hero) {
    // check whether the hero is still present
    if (!test_collision_inside(get_hero())) {
      activated_by_hero = false;

      // Notify Lua.
      notifying_script = true;
      get_lua_context()->sensor_on_left(*this);
      notifying_script = false;
    }
  }
}

}

File: tools/org/solarus/editor/entities/Sensor.java
/*
 * Copyright (C) 2006-2012 Christopho, Solarus - http://www.solarus-games.org
 *
 * Solarus Quest Editor is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Zelda: Mystery of Solarus DX is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */
package org.solarus.editor.entities;

import org.solarus.editor.*;
import java.awt.*;

/**
 * A sensor is a place that triggers something when the hero
 * or other entities walk on it.
 * Unlike switches and teletransportes, the hero cannot avoid a sensor
 * when he is jumping.
 */
public class Sensor extends MapEntity {

    /**
     * Origin point of this entity type.
     */
    private static final Point origin = new Point(8, 13);

    /**
     * Unitary size of this entity type.
     */
    private static final Dimension unitarySize = new Dimension(16, 16);

    /**
     * Resizable image of this entity type.
     */
    private static Image resizableImage;

    /**
     * Description of the default image representing this kind of entity.
     */
    public static final EntityImageDescription[] generalImageDescriptions = {
        new EntityImageDescription("sensors.png", 0, 0, 32, 32)
    };

    /**
     * Creates a new switch.
     * @param map the map
     */
    public Sensor(Map map) throws MapException {
        super(map, 16, 16);
    }

    /**
     * Returns whether the entity has an identifier.
     * @return true
     */
    public boolean hasName() {
        return true;
    }

    /**
     * Returns whether or not the entity is currently resizable.
     * @return true
     */
    public boolean isResizable() {
        return true;
    }

    /**
     * Returns whether two entities of this type can have different
     * sizes, even if they are not resizable.
     * @return true
     */
    public boolean isSizeVariable() {
        return true;
    }

    /**
     * Returns the minimum size of the entity (for a resizable entity).
     * When the entity is resized, its new size must be a multiple of this minimum size.
     * @return (16,16)
     */
    public Dimension getUnitarySize() {
        return unitarySize;
    }

    /**
     * Returns the coordinates of the origin point of the entity.
     * @return (8,13)
     */
    protected Point getOrigin() {
        return origin;
    }

    /**
     * Draws the entity on the map editor.
     * @param g graphic context
     * @param zoom zoom of the image (for example, 1: unchanged, 2: zoom of 200%)
     * @param showTransparency true to make transparent pixels,
     * false to replace them by a background color
     */
    public void paint(Graphics g, double zoom, boolean showTransparency) {

        if (resizableImage == null) {
            resizableImage = Project.getEditorImage("resizable_sensors.png");
        }

        int x = (int) (positionInMap.x * zoom);
        int y = (int) (positionInMap.y * zoom);
        int w = (int) (positionInMap.width * zoom);
        int h = (int) (positionInMap.height * zoom);

        g.setColor(new Color(64, 160, 64));
        g.fillRect(x, y, w, h);

        int dx1 = (int) ((positionInMap.x + positionInMap.width / 2 - 8) * zoom);
        int dy1 = (int) ((positionInMap.y + positionInMap.height / 2 - 8) * zoom);
        int dx2 = (int) (dx1 + 16 * zoom);
        int dy2 = (int) (dy1 + 16 * zoom);

        int sx1 = 32 * getSubtypeId();
        int sx2 = sx1 + 32;

        g.drawImage(resizableImage, dx1, dy1, dx2, dy2, sx1, 0, sx2, 32, null);

        drawEntityOutline(g, zoom, new Color(142, 240, 142));
    }
}

